test/e2e/apimachinery/crd_validation_ratcheting.go

```diff
@@ -0,0 +1,898 @@
+			field: "notfoo"
+			struct:
+				field: "notfoo"
+			list:
+			- key: "foo"
+			  field: "notfoo"
+			- key: "bar"
+			  field: "notfoo"
+			map:
+				foo:
+					field: "notfoo"
+				bar:
+					field: "notfoo"
+		`)
+		framework.ExpectNoError(err, "parsing modified resource")
+		err = applyPatch(testCRDGVR, "test-resource", modifiedInstance.Object)
+		for _, fieldPath := range []string{
+			"field",
+			"struct.field",
+			"list[0].field",
+			"list[1].field",
+			"map[foo].field",
+			"map[bar].field",
+		} {
+			gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring(fieldPath)))
+		}
+	})
+
+	ginkgo.It("MUST NOT ratchet errors raised by transition rules", func() {
+		ginkgo.By("creating an initial object with many correlatable fields")
+		instance, err := parseUnstructured(`
+			field: "foo"
+			struct:
+				field: "foo"
+			list:
+			- key: "foo"
+			  field: "foo"
+			- key: "bar"
+			  field: "foo"
+			map:
+				foo:
+					field: "foo"
+				bar:
+					field: "foo"
+		`)
+		framework.ExpectNoError(err, "parsing test resource")
+		framework.ExpectNoError(applyPatch(testCRDGVR, "test-resource", instance.Object), "failed creating test resource")
+
+		ginkgo.By("updating CRD schema with constraints on correlatable fields to make instance invalid")
+		sch, err := parseSchema(`
+			type: object
+			properties:
+				field: 
+					type: string
+					maxLength: 5
+					x-kubernetes-validations:
+					- rule: self != oldSelf
+				struct:
+					type: object
+					properties:
+						field:
+							type: string
+							maxLength: 5
+							x-kubernetes-validations:
+							- rule: self != oldSelf
+				list:
+					type: array
+					maxItems: 5
+					x-kubernetes-list-type: map
+					x-kubernetes-list-map-keys: [key]
+					items:
+						type: object
+						properties:
+							key: {type: string}
+							field:
+								type: string
+								maxLength: 5
+								x-kubernetes-validations:
+								- rule: self != oldSelf
+						required:
+						- key
+				map:
+					type: object
+					maxProperties: 5
+					additionalProperties:
+						type: object
+						properties:
+							field:
+								type: string
+								maxLength: 5
+								x-kubernetes-validations:
+								- rule: self != oldSelf
+		`)
+
+		framework.ExpectNoError(err, "parsing schema")
+		framework.ExpectNoError(updateCRDSchema(testCRDGVR, *sch), "failed to update schema")
+
+		ginkgo.By("updating a label on the test resource")
+		instance.SetLabels(map[string]string{
+			"foo": "bar",
+		})
+		err = applyPatch(testCRDGVR, "test-resource", instance.Object)
+		for _, fieldPath := range []string{
+			"field",
+			"struct.field",
+			"list[0].field",
+			"list[1].field",
+			"map[foo].field",
+			"map[bar].field",
+		} {
+			gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring(fieldPath)))
+		}
+	})
+
+	// ginkgo.It("MUST evaluate a CRD Validation Rule with oldSelf = nil for new values when optionalOldSelf is true", func() {
+	// })
```

Comment:
Can we replace this with some test that levarages optionals? That's the only think I  see missing in these new tests.

Reply:
Added